File: a1z/dynamics/gravity_model.py
"""Pinocchio RNEA-based gravity compensation model."""

import os

import numpy as np

try:
    import pinocchio
except ImportError:
    raise ImportError("Pinocchio is required. Install with: pip install pin")


class GravityModel:
    """Compute gravity compensation torques from a URDF using Pinocchio RNEA."""

    def __init__(self, urdf_path: str, mesh_dir: str = ""):
        """Load URDF and build Pinocchio model.

        Args:
            urdf_path: Path to the URDF file.
            mesh_dir: Root directory for mesh resources (optional, not needed
                      for pure dynamics computation).
        """
        urdf_path = os.path.expanduser(urdf_path)
        if not os.path.isfile(urdf_path):
            raise FileNotFoundError(f"URDF not found: {urdf_path}")

        self.model = pinocchio.buildModelFromUrdf(urdf_path)
        self.data = self.model.createData()
        self.nq = self.model.nq
        self.nv = self.model.nv

        assert np.allclose(self.model.gravity.linear, [0, 0, -9.81]), (
            f"Unexpected gravity vector: {self.model.gravity.linear}. "
            "The base_link Z axis should point upward."
        )

    def compute_gravity_torque(self, q: np.ndarray) -> np.ndarray:
        """Compute gravity compensation torque tau_g(q).

        Uses RNEA with zero velocity and acceleration, yielding the static
        gravity term g(q).

        Args:
            q: Joint angles (rad), shape (nq,).

        Returns:
            tau_g: Gravity torques (Nm), shape (nv,).
        """
        assert q.shape == (self.nq,), f"Expected q shape ({self.nq},), got {q.shape}"
        zero_v = np.zeros(self.nv)
        tau_g = pinocchio.rnea(self.model, self.data, q, zero_v, zero_v)
        return tau_g.copy()

    def compute_inverse_dynamics(
        self, q: np.ndarray, qd: np.ndarray, qdd: np.ndarray
    ) -> np.ndarray:
        """Compute full inverse dynamics torque via RNEA.

        tau = M(q)*qdd + C(q,qd)*qd + g(q)

        When qd=0 and qdd=0 this degenerates to compute_gravity_torque.

        Args:
            q:   Joint positions (rad), shape (nq,).
            qd:  Joint velocities (rad/s), shape (nv,).
            qdd: Joint accelerations (rad/s^2), shape (nv,).

        Returns:
            tau: Inverse dynamics torques (Nm), shape (nv,).
        """
        assert q.shape == (self.nq,), f"Expected q shape ({self.nq},), got {q.shape}"
        assert qd.shape == (self.nv,), f"Expected qd shape ({self.nv},), got {qd.shape}"
        assert qdd.shape == (self.nv,), f"Expected qdd shape ({self.nv},), got {qdd.shape}"
        tau = pinocchio.rnea(self.model, self.data, q, qd, qdd)
        return tau.copy()

    def get_joint_limits(self) -> np.ndarray:
        """Return joint limits array, shape (nq, 2): [[lower, upper], ...]."""
        lower = self.model.lowerPositionLimit
        upper = self.model.upperPositionLimit
        return np.column_stack([lower, upper])

    def set_gravity(self, gravity: np.ndarray) -> None:
        """Set gravity vector (for non-standard mounting orientations)."""
        assert gravity.shape == (3,)
        self.model.gravity = pinocchio.Motion(gravity, np.zeros(3))

File: a1z_cpp/src/gravity_model.cpp
#include "a1z/gravity_model.hpp"

#include <cmath>
#include <iostream>
#include <stdexcept>

namespace a1z {

GravityModel::GravityModel(const std::string& urdf_path, const std::string& mesh_dir) {
#ifdef A1Z_HAS_PINOCCHIO
    try {
        pinocchio::urdf::buildModel(urdf_path, model_);
        data_ = pinocchio::Data(model_);
        nq_ = model_.nq;
        nv_ = model_.nv;
        has_model_ = true;

        // Verify gravity direction
        const auto& g = model_.gravity.linear();
        if (std::abs(g[0]) > 1e-6 || std::abs(g[1]) > 1e-6 || g[2] > 0) {
            std::cerr << "[GravityModel] Warning: unexpected gravity vector: ["
                      << g[0] << ", " << g[1] << ", " << g[2] << "]" << std::endl;
        }
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to load URDF: " + std::string(e.what()));
    }
#else
    (void)urdf_path;
    (void)mesh_dir;
    std::cerr << "[GravityModel] Warning: Pinocchio not available, gravity compensation disabled"
              << std::endl;
    has_model_ = false;
#endif
}

JointVector GravityModel::compute_gravity_torque(const JointVector& q) {
    JointVector zero = {};
    return compute_inverse_dynamics(q, zero, zero);
}

JointVector GravityModel::compute_inverse_dynamics(const JointVector& q,
                                                    const JointVector& qd,
                                                    const JointVector& qdd) {
    JointVector tau = {};

#ifdef A1Z_HAS_PINOCCHIO
    if (!has_model_) {
        return tau;
    }

    Eigen::VectorXd q_eigen(nq_), qd_eigen(nv_), qdd_eigen(nv_);
    for (int i = 0; i < nq_; ++i) q_eigen[i] = q[i];
    for (int i = 0; i < nv_; ++i) {
        qd_eigen[i] = qd[i];
        qdd_eigen[i] = qdd[i];
    }

    Eigen::VectorXd tau_eigen = pinocchio::rnea(model_, data_, q_eigen, qd_eigen, qdd_eigen);

    for (int i = 0; i < nv_ && i < 6; ++i) {
        tau[i] = tau_eigen[i];
    }
#else
    (void)q;
    (void)qd;
    (void)qdd;
#endif

    return tau;
}

std::array<std::pair<double, double>, 6> GravityModel::get_joint_limits() const {
    std::array<std::pair<double, double>, 6> limits;

#ifdef A1Z_HAS_PINOCCHIO
    if (has_model_) {
        for (int i = 0; i < nq_ && i < 6; ++i) {
            limits[i] = {model_.lowerPositionLimit[i], model_.upperPositionLimit[i]};
        }
        return limits;
    }
#endif

    // Default A1Z limits
    limits[0] = {-2.094, 2.094};   // J1
    limits[1] = {0.0, 3.142};      // J2
    limits[2] = {-3.142, 0.0};     // J3
    limits[3] = {-1.484, 1.484};   // J4
    limits[4] = {-1.484, 1.484};   // J5
    limits[5] = {-2.007, 2.007};   // J6
    return limits;
}

void GravityModel::set_gravity(const std::array<double, 3>& gravity) {
#ifdef A1Z_HAS_PINOCCHIO
    if (has_model_) {
        model_.gravity.linear(Eigen::Vector3d(gravity[0], gravity[1], gravity[2]));
    }
#else
    (void)gravity;
#endif
}

bool GravityModel::is_available() {
#ifdef A1Z_HAS_PINOCCHIO
    return true;
#else
    return false;
#endif
}

} // namespace a1z
